### PR TITLE
Throw Anonymous Child of AnalysisException [databricks]

### PR DESCRIPTION
**Pull request by razajafri**
In Spark 4.0.0 the constructor of AnalysisException that takes a String parameter has been protected. One way to get around this is to add a shim where we can throw a defined class for Spark 4.0.0 while keeping the same logic for other versions of Spark. That would be a lot of code churn and be hard to review for no added benefit IMO.

This change is small enough and simple enough to where we can throw an anonymous class without making it difficult to read

contributes to #9259

### CODE REVIEW

File: sql-plugin/src/main/scala/org/apache/spark/sql/hive/rapids/RapidsHiveErrors.scala [hunk 1]
```diff
@@ -54,7 +54,7 @@ object RapidsHiveErrors {
 
   def cannotResolveAttributeError(name: String, outputStr: String): Throwable = {
     new AnalysisException(
-      s"Unable to resolve $name given [$outputStr]")
+      s"Unable to resolve $name given [$outputStr]") {}
```
Comment (jlowe) [inline sql-plugin/src/main/scala/org/apache/spark/sql/hive/rapids/RapidsHiveErrors.scala:57]: I don't think this is a good approach, as it's very unfriendly to the end-user.  The exception is going to show up without AnalysisException in the name -- it will be something like RapidsHiveErrors$$anon$## (or whatever the enclosing class is where the anonymous class was created), and thus won't look at all like the AnalysisException class that the user is used to seeing.  Yes, it's semantically like an AnalysisException since it derives from it, but the way it appears to the end user in a stacktrace does not look similar at all.

The reason Spark made the string private is because they want the code to go through QueryExecutionErrors and other similar error reporting interfaces.  We have our own wrappers to that in shims, so maybe the best way to fix this is to migrate these use cases to those already shim'd error interfaces, and have the older shims call AnalysisException and the new shims use the newer error reporting that doesn't use an arbitrary string.  If we really want to match the Spark behavior here, that's the right way to do it.

Comment (razajafri) [reply sql-plugin/src/main/scala/org/apache/spark/sql/hive/rapids/RapidsHiveErrors.scala:57]: I agree, that's the right way to do it and went down that path but I wasn't sure if we will be OK with the difficulty in code review and decided to post this before going back that route. Will update in a few

**Comment by razajafri** [comment]
@jlowe I think I have addressed all your concerns

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDataWritingCommandExec.scala [hunk 2]
```diff
@@ -84,10 +84,8 @@ object GpuDataWritingCommand {
       if (fs.exists(filePath) &&
           fs.getFileStatus(filePath).isDirectory &&
           fs.listStatus(filePath).length != 0) {
-        TrampolineUtil.throwAnalysisException(
-          s"CREATE-TABLE-AS-SELECT cannot create table with location to a non-empty directory " +
-              s"${tablePath} . To allow overwriting the existing non-empty directory, " +
-              s"set '$allowNonEmptyLocationInCTASKey' to true.")
+        AnalysisExceptionShim.throwException("_LEGACY_ERROR_TEMP_1241",
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDataWritingCommandExec.scala:87]: We should not be throwing literal magic strings.  This is really fragile and likely to have a typo that won't be caught by the compiler and thus likely to not do what we want at runtime.  The Spark code is not doing this in all the various places in Catalyst, instead those are calling QueryExecutionErrors with a descriptive method name, and _that_ is doing the specifics of throwing a specific message.

That's what I mentioned before:
> We have our own wrappers to that in shims, so maybe the best way to fix this is to migrate these use cases to those already shim'd error interfaces

We shouldn't be creating a new shim class here, but rather using the one we already have, i.e.: RapidsErrorUtils.  For example, for this case here we should add a `createTableAsSelectWithNonEmptyDirectoryError` method to the RapidsErrorUtils shim class.  On modern Spark versions, it would simply delegate to Spark's QueryCompilationErrors.  On older Spark versions, the RapidsErrorUtils shim would throw an AnalysisException with an appropriate message string it constructs based on the arguments.

In short, see what the new Spark code is doing in these areas.  I suspect none of them are throwing AnalysisException but rather calling an error utility class like QueryExecutionErrors or QueryCompilationErrors.  We should call the same methods on Spark versions that have those methods available, not throw a derived class.

Comment (razajafri) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDataWritingCommandExec.scala:87]: Ah, I was fixated on solving it using the errorClass even for older versions of Spark.

Comment (jlowe) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDataWritingCommandExec.scala:87]: Yeah the shim logic should almost always be "do what this version of Spark does for this" whether that's using an error class abstraction or throwing a raw AnalysisException with a string message.

Comment (razajafri) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDataWritingCommandExec.scala]: OK. I think I got it right this time. This will be painful to review PTAL

File: sql-plugin/src/main/spark311/scala/org/apache/spark/sql/rapids/shims/RapidsQueryErrorUtils.scala [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+package org.apache.spark.sql.rapids.shims
+
+import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.types.StructType
+
+trait RapidsQueryErrorUtils {
+
+  def createTableAsSelectWithNonEmptyDirectoryError(tablePath: String, conf: String): Throwable = {
+    new AnalysisException(s"CREATE-TABLE-AS-SELECT cannot create table with location to a " +
```
Comment (jlowe) [inline sql-plugin/src/main/spark311/scala/org/apache/spark/sql/rapids/shims/RapidsQueryErrorUtils.scala:53]: Many (all?) of these should be leveraging either QueryExecutionErrors or QueryCompilationErrors rather than throwing raw analysis exceptions on the Spark versions that support those classes, just like the Spark 4 version of this shim does.  Not sure it makes sense to isolate these as a trait from the other methods in RapidsErrorUtils.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuInsertIntoHadoopFsRelationCommand.scala [hunk 4]
```diff
@@ -121,7 +122,7 @@ case class GpuInsertIntoHadoopFsRelationCommand(
       val pathExists = fs.exists(qualifiedOutputPath)
       (mode, pathExists) match {
         case (SaveMode.ErrorIfExists, true) =>
-          throw new AnalysisException(s"path $qualifiedOutputPath already exists.")
+          TrampolineUtil.throwAnalysisException(s"path $qualifiedOutputPath already exists.")
```
[the remark this answers is not in the document]
Comment (razajafri) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuInsertIntoHadoopFsRelationCommand.scala:125]: I saw that and decided not to use it because it's adding `Set mode as "overwrite" to overwrite the existing path.` which we didn't have in our original message.

[the remark this answers is not in the document]
Comment (jlowe) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuInsertIntoHadoopFsRelationCommand.scala:125]: But that extra message is applicable here, and we would want that.  The message was updated in Spark and we didn't update accordingly.

File: sql-plugin/src/main/spark311/scala/org/apache/spark/sql/rapids/shims/RapidsQueryErrorUtils.scala (hunk 3 above)
Comment (razajafri) [reply sql-plugin/src/main/spark311/scala/org/apache/spark/sql/rapids/shims/RapidsQueryErrorUtils.scala:37]: I have moved the shim to Spark 320 which is  the version when the relevant `QueryCompilationErrors` were introduced

File: sql-plugin/src/main/spark340/scala/org/apache/spark/sql/rapids/shims/RapidsErrorUtilsFor340Plus.scala [hunk 5]
```diff
@@ -0,0 +1,95 @@
+/*
```
Comment (razajafri) [inline sql-plugin/src/main/spark340/scala/org/apache/spark/sql/rapids/shims/RapidsErrorUtilsFor340Plus.scala:1]: Renamed the old `RapidsErrorUtils`

**Comment by razajafri** [comment]
I have addressed all your concerns I think PTAL. 
The only thing that you may have questions about is that I have introduced the RapidsAnalysisException which will be thrown even in older versions of Spark instead of Raw AnalysisException

**Comment by razajafri** [comment]
build

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuInsertIntoHadoopFsRelationCommand.scala [hunk 6]
```diff
@@ -121,7 +121,7 @@ case class GpuInsertIntoHadoopFsRelationCommand(
       val pathExists = fs.exists(qualifiedOutputPath)
       (mode, pathExists) match {
         case (SaveMode.ErrorIfExists, true) =>
-          throw new AnalysisException(s"path $qualifiedOutputPath already exists.")
+          throw RapidsErrorUtils.dataPathNotExistError(qualifiedOutputPath.toString)
```
Comment (jlowe) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuInsertIntoHadoopFsRelationCommand.scala:124]: This is the wrong error message to throw for this error.  The error message will state that it does _not_ exist, but instead it _does_exist, and that's the problem.  RapidsQueryErrorUtils.outputPathAlreadyExistsError would be more appropriate.  Note that the error being thrown here in Spark 4.0.0 is via QueryExecutionErrors.saveModeUnsupportedError.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/catalyst/expressions/GpuRandomExpressions.scala [hunk 7]
```diff
@@ -52,7 +52,7 @@ case class GpuRand(child: Expression) extends ShimUnaryExpression with GpuExpres
   @transient protected lazy val seed: Long = child match {
     case GpuLiteral(s, IntegerType) => s.asInstanceOf[Int]
     case GpuLiteral(s, LongType) => s.asInstanceOf[Long]
-    case _ => throw new AnalysisException(
+    case _ => TrampolineUtil.throwAnalysisException(
```
Comment (jlowe) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/catalyst/expressions/GpuRandomExpressions.scala:55]: Why do we have a Trampoline method for throwing an AnalysisException _and_ a RapidsAnalysisException?  Seems like we should only have one of these, not both (probably just the latter).

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/TrampolineUtil.scala [hunk 8]
```diff
@@ -241,4 +241,11 @@ object TrampolineUtil {
   }
 
   def getSparkHadoopUtilConf: Configuration = SparkHadoopUtil.get.conf
+
 }
+
+/**
+ * This class is to only be used when there is no equivalent helper method in
+ * Spark to throw an AnalysisException.
```
Comment (jlowe) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/TrampolineUtil.scala:249]:
```suggestion
 * This class is to only be used to throw errors specific to the
 * RAPIDS Accelerator or errors mirroring Spark where a raw
 * AnalysisException is thrown directly rather than via an error
 * utility class (this should be rare).
```

File: sql-plugin/src/main/spark332db/scala/org/apache/spark/sql/rapids/GpuFileFormatWriter.scala [hunk 9]
```diff
@@ -61,7 +62,7 @@ object GpuFileFormatWriter extends Logging {
   private def verifySchema(format: ColumnarFileFormat, schema: StructType): Unit = {
     schema.foreach { field =>
       if (!format.supportDataType(field.dataType)) {
-        throw new AnalysisException(
+        TrampolineUtil.throwAnalysisException(
```
Comment (jlowe) [inline sql-plugin/src/main/spark332db/scala/org/apache/spark/sql/rapids/GpuFileFormatWriter.scala:65]: Should use RapidsAnalysisException here.

File: sql-plugin/src/main/spark400/scala/org/apache/spark/sql/rapids/shims/RapidsErrorUtils.scala [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+spark-rapids-shim-json-lines ***/
+package org.apache.spark.sql.rapids.shims
+
+import org.apache.spark.sql.AnalysisException
+
+object RapidsErrorUtils extends RapidsErrorUtilsFor340Plus {
+  override def dynamicPartitionParentError: Throwable = {
+    /** This exception doesn't have a helper method so the errorClass has to be hardcoded */
+    throw new AnalysisException(
+      errorClass = "_LEGACY_ERROR_TEMP_3079",
```
Comment (jlowe) [inline sql-plugin/src/main/spark400/scala/org/apache/spark/sql/rapids/shims/RapidsErrorUtils.scala:28]: I'm worried about this changing out from underneath us and then the error doesn't work at all.  The name itself implies this is temporary and will be changing, yet the compiler won't catch when that occurs.  It might be better to throw an error message directly here for now rather than rely on this tenuous connection to Spark's error message.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuInsertIntoHadoopFsRelationCommand.scala (hunk 6 above)
Comment (razajafri) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuInsertIntoHadoopFsRelationCommand.scala:124]: Thanks. 

> Note that the error being thrown here in Spark 4.0.0 is via QueryExecutionErrors.saveModeUnsupportedError.

I am not sure what you mean here.

Comment (jlowe) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuInsertIntoHadoopFsRelationCommand.scala:124]: Just noting that there's a new interface in Spark 4.0.0 for throwing this type of error, see https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala#L139.  I think it's probably fine to use the outputPathAlreadyExistsError, but technically Spark 4.0.0 is using a different error utility method here.

File: sql-plugin/src/main/spark400/scala/org/apache/spark/sql/rapids/shims/RapidsErrorUtils.scala (hunk 10 above)
Comment (razajafri) [reply sql-plugin/src/main/spark400/scala/org/apache/spark/sql/rapids/shims/RapidsErrorUtils.scala]: I have replaced this with the current way of throwing `ErrorMsg.PARTITION_DYN_STA_ORDER.getMsg`.

**Comment by razajafri** [comment]
I have addressed your comments I think. 
* Renamed the method in `TrampolineUtil.throwAnalysisException()` to `TrampolineUtil.throwRapidsAnalysisException` 
* Throwing the ErrorMsg.PARTITION_DYN_STA_ORDER.getMsg in dynamicPartitionParentError instead of the errorClass
* Removed the `RapidsErrorUtilsFor340PlusShims`

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/TrampolineUtil.scala [hunk 11]
```diff
@@ -157,8 +157,14 @@ object TrampolineUtil {
...
-  /** Throw a Spark analysis exception */
-  def throwAnalysisException(msg: String) = throw new AnalysisException(msg)
+  /**
+   * Throw a Rapids analysis exception
+   * This method is to only be used to throw errors specific to the
+   * RAPIDS Accelerator or errors mirroring Spark where a raw
+   * AnalysisException is thrown directly rather than via an error
+   * utility class (this should be rare).
+   */
+  def throwRapidsAnalysisException(msg: String) = throw new RapidsAnalysisException(msg)
```
[the remark this answers is not in the document]
Comment (razajafri) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/TrampolineUtil.scala:167]: We don't need a `TrampolineUtil` method, this was a remnant of throwing the AnalysisException.

[the remark this answers is not in the document]
Comment (razajafri) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/TrampolineUtil.scala:167]: Let me get rid of that. Please let me know if there are any other concerns

**Comment by razajafri** [comment]
build

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/catalyst/expressions/GpuRandomExpressions.scala (hunk 7 above)
Comment (mythrocks) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/catalyst/expressions/GpuRandomExpressions.scala]: Looks like this was resolved by switching to RapidsAnalysisException.

**Review by mythrocks — APPROVED** [review]
I have verified that all of @jlowe's concerns were addressed here.

LGTM!
